tsl/src/continuous_aggs/refresh.c

```diff
@@ -220,21 +152,71 @@ continuous_agg_refresh_execute(const CaggRefreshState *refresh)
 	};
 	Dimension *time_dim = hyperspace_get_open_dimension(refresh->cagg_ht->space, 0);
 
+	Assert(time_dim != NULL);
+
 	continuous_agg_update_materialization(refresh->partial_view,
 										  cagg_hypertable_name,
 										  &time_dim->fd.column_name,
-										  refresh->refresh_window,
+										  *bucketed_refresh_window,
 										  unused_invalidation_range,
 										  refresh->cagg.data.bucket_width);
 }
 
 static void
-continuous_agg_refresh_with_window(ContinuousAgg *cagg, const InternalTimeRange *refresh_window)
+continuous_agg_refresh_with_window(ContinuousAgg *cagg, const InternalTimeRange *refresh_window,
+								   InvalidationStore *invalidations)
 {
 	CaggRefreshState refresh;
+	TupleTableSlot *slot;
 
 	continuous_agg_refresh_init(&refresh, cagg, refresh_window);
-	continuous_agg_refresh_execute(&refresh);
+	slot = MakeSingleTupleTableSlotCompat(invalidations->tupdesc, &TTSOpsMinimalTuple);
+
+	while (tuplestore_gettupleslot(invalidations->tupstore,
+								   true /* forward */,
+								   false /* copy */,
+								   slot))
+	{
+		bool isnull;
+		Datum start = slot_getattr(
+			slot,
+			Anum_continuous_aggs_materialization_invalidation_log_lowest_modified_value,
+			&isnull);
+		Datum end = slot_getattr(
+			slot,
+			Anum_continuous_aggs_materialization_invalidation_log_greatest_modified_value,
+			&isnull);
+		InternalTimeRange invalidation = {
+			.type = refresh_window->type,
+			.start = DatumGetInt64(start),
+			.end = DatumGetInt64(end),
+		};
+		InternalTimeRange bucketed_refresh_window =
+			compute_bucketed_refresh_window(&invalidation, cagg->data.bucket_width);
+
+		if (client_min_messages <= DEBUG1)
+		{
```

Comment:
This code is not covered at all, so we have no idea if it can cause crashes or some other problems. Just having a test that run the code (even if it throws away the output) would be a great help in avoiding an embarrassing crashing bug here.